src/raglite/_rag.py

```diff
@@ -2,148 +2,95 @@
-        context_neighbors=context_neighbors,
-        messages=messages,
-        config=config,
-    )
-    # Retrieve the top chunks.
+) -> list[ChunkSpan]:
+    """Retrieve context for RAG."""
+    # If the user has configured a reranker, we retrieve extra contexts to rerank.
     config = config or RAGLiteConfig()
-    chunks: list[str] | list[Chunk]
-    if callable(search):
-        # If the user has configured a reranker, we retrieve extra contexts to rerank.
-        extra_contexts = 3 * max_contexts if config.reranker else 0
-        # Retrieve relevant contexts.
-        chunk_ids, _ = search(prompt, num_results=max_contexts + extra_contexts, config=config)
-        # Rerank the relevant contexts.
-        chunks = rerank_chunks(query=prompt, chunk_ids=chunk_ids, config=config)
-    else:
-        # The user has passed a list of chunk_ids or chunks directly.
-        chunks = search
+    extra_chunks = 3 * num_chunks if config.reranker else 0
+    # Search for relevant chunks.
+    chunk_ids, _ = search(query, num_results=num_chunks + extra_chunks, config=config)
+    # Rerank the chunks from most to least relevant.
+    chunks = rerank_chunks(query, chunk_ids=chunk_ids, config=config)
     # Extend the top contexts with their neighbors and group chunks into contiguous segments.
-    segments = retrieve_segments(chunks[:max_contexts], neighbors=context_neighbors, config=config)
-    return segments
+    context = retrieve_chunk_spans(chunks[:num_chunks], neighbors=chunk_neighbors, config=config)
+    return context
 
 
-def generate(
-    prompt: str,
+def create_rag_instruction(
+    user_prompt: str,
+    context: list[ChunkSpan],
     *,
-    system_prompt: str = RAG_SYSTEM_PROMPT,
-    messages: list[dict[str, str]] | None = None,
-    context_segments: list[ContextSegment],
-    config: RAGLiteConfig,
-) -> Iterator[str]:
-    messages = _compose_messages(
-        prompt=prompt, system_prompt=system_prompt, messages=messages, segments=context_segments
-    )
-    # Stream the LLM response.
-    stream = completion(
-        model=config.llm,
-        messages=_compose_messages(
-            prompt=prompt, system_prompt=system_prompt, messages=messages, segments=context_segments
+    rag_instruction_template: str = RAG_INSTRUCTION_TEMPLATE,
+) -> dict[str, str]:
+    """Convert a user prompt to a RAG instruction.
+
+    The RAG instruction's format follows Anthropic's best practices [1].
+
+    [1] https://docs.anthropic.com/en/docs/build-with-claude/prompt-engineering/long-context-tips
+    """
+    message = {
+        "role": "user",
+        "content": rag_instruction_template.format(
+            user_prompt=user_prompt,
+            context="\n".join(
+                chunk_span.to_xml(index=i + 1) for i, chunk_span in enumerate(context)
+            ),
         ),
-        stream=True,
-    )
+    }
+    return message
+
+
+def rag(messages: list[dict[str, str]], *, config: RAGLiteConfig) -> Iterator[str]:
```

Comment:
Where is the System prompt? Is it supposed to be in messages? Not sure about it as it will force developers to add it by hand.

Reply:
My reasoning was that if we completely follow Anthropic's long-context best practices (which does not rely on a system prompt), we actually catch two (three!) birds with one stone because that approach is also optimal for prompt caching. And the third bird is that it allows to easily truncate the number of messages in the message history to what fits in the LLM's context window without having to worry about dropping our system prompt.

Reply:
On the other hand, I like that we give flexibility of adding the context explicitly wherever the user considers it more convenient, so I am divided.